Migrate Students-List Component

HasinduWelarathne opened the pull request:
# Description

This PR migrates the existing students-list component from AngularJS/CoffeeScript to Angular 17/TypeScript, adopting Angular Material and Tailwind CSS. All original functionality (search/typeahead, filters, sorting, progress bars, flag icons, campus/tutorial selects, CSV export, enrol modal, pagination, and row navigation) remains behaviorally identical.

## Component Review  
https://github.com/thoth-tech/documentation/pull/597

## Type of change  
Migration

# How Has This Been Tested?

- Verified typing in the search bar filters the list and shows autocomplete suggestions.  
- Toggled “Advanced Filters” and applied “All Tutorials” vs “My Tutorials” filters.  
- Clicked each sort button (Grade, Plagiarism, Portfolio) twice to confirm original and reverse ordering.  
- Entered a term yielding no matches and saw the “No students found” alert.  
- Scanned table rows to confirm avatars, usernames, names, progress bars, and flag icons display correctly.  
- Opened campus and tutorial dropdowns to change selections and confirmed the list updates.  
- Clicked on a student row or avatar and confirmed navigation to the student details view.  
- Clicked “Export CSV” and verified the downloaded file’s headers and rows match the table.  
- Clicked “Enrol Student” and ensured the enrolment modal opens.  
- Used pagination controls to switch pages and change page size, ensuring the list updates.

## Screenshots  
**Before:**  
<img width="1913" height="912" alt="image" src="https://github.com/user-attachments/assets/d944bbf2-b7e5-4eb7-a05f-beb2012068ba" />

**After:**  
<img width="1919" height="921" alt="image" src="https://github.com/user-attachments/assets/62715373-3c07-4a86-891a-f467e786e509" />

# Testing Checklist

- [x] Tested in latest Chrome  
- [x] Tested in latest Safari  
- [x] Tested in latest Firefox  

# Checklist

- [x] My code follows the style guidelines of this project  
- [x] I have performed a self-review of my own code  
- [x] I have commented my code in hard-to-understand areas  
- [x] I have made corresponding changes to the documentation  
- [x] My changes generate no new warnings

HasinduWelarathne committed:
WIP:complete migration to Angular 17 + Tailwind

HasinduWelarathne committed:
Migrate students-list component

HasinduWelarathne committed:
fix: update student-list component template and styles

HasinduWelarathne committed:
style(students-list): unify tab selector styling and disable flag sorting

HasinduWelarathne committed:
fix: migrate students-list.component.ts to Angular 17

Pasindufdo98 reviewed (commented):
Hi Hasindu, 
I pulled your branch and tested locally, I was able to successfully build and run the application. When I was interacting with your component, I found that students are not correctly mapping when selecting a student in the list.
As an example, 
when I selected student 1, it navigates to student 8 so on.
 And rest of the functionalities looks good to me.

<img width="1920" height="1133" alt="Screenshot (1588)" src="https://github.com/user-attachments/assets/0cc2568e-2a5d-45eb-8a82-47d12eeb5fbe" />

HasinduWelarathne committed:
fix(students-list): use project.id instead of student.id for navigation

HasinduWelarathne commented:
> Hi Hasindu, I pulled your branch and tested locally, I was able to successfully build and run the application. When I was interacting with your component, I found that students are not correctly mapping when selecting a student in the list. As an example, when I selected student 1, it navigates to student 8 so on. And rest of the functionalities looks good to me.
> 
> <img alt="Screenshot (1588)" width="1920" height="1133" src="https://private-user-images.githubusercontent.com/116358471/478905193-0cc2568e-2a5d-45eb-8a82-47d12eeb5fbe.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU1MjE5MzEsIm5iZiI6MTc1NTUyMTYzMSwicGF0aCI6Ii8xMTYzNTg0NzEvNDc4OTA1MTkzLTBjYzI1NjhlLTJhNWQtNDVlYi04YTgyLTQ3ZDEyZWViNWZiZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgxOFQxMjUzNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMjYwNzNjMWQ0NmNjMDljN2ZkNGNkMzYyZjFkMjQwZWJkNjUxMDEyZGE1Y2VjNDBhYTZmYmU4MTEzYjk0ZGU2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.a3zxP4he9Yl87nSQvRj0Wk5XqQPFR5CThuxriT9Wqig">

Hi @Pasindufdo98 ,

Thanks a lot for testing and pointing that out. You were absolutely right, the navigation was using student.id instead of the project’s id, which caused the mismatch. I’ve fixed it now so that the correct student opens when selecting a student.
Appreciate your thorough review!

WaelAlahamdi commented:
Hi Hasindu,
I was able to build and run the branch successfully. The new student list view and navigation to individual student dashboards are working as intended.
One issue I observed:
•	When selecting a student and then navigating back to the student list, the Stats progress bar in the list disappears.  See Student_10 in the screenshot to clarify.
•	The progress data only reloads after a full page refresh, which may affect usability for tutors monitoring students.
Overall, the functionality looks good, but this refresh issue should be addressed to ensure consistent display of student stats.
<img width="1375" height="857" alt="Feedback PR #389" src="https://github.com/user-attachments/assets/0cc25cc7-38b4-4d3a-a554-562efcbfa206" />

HasinduWelarathne committed:
fix(students-list): restore progress stats after navigating back from dashboard

- Added sessionStorage-based snapshot mechanism to restore student's taskStats.
- Ensures correct stats display when returning from student dashboard.
- Keeps other component behavior unchanged.

HasinduWelarathne commented:
> Hi Hasindu, I was able to build and run the branch successfully. The new student list view and navigation to individual student dashboards are working as intended. One issue I observed: • When selecting a student and then navigating back to the student list, the Stats progress bar in the list disappears. See Student_10 in the screenshot to clarify. • The progress data only reloads after a full page refresh, which may affect usability for tutors monitoring students. Overall, the functionality looks good, but this refresh issue should be addressed to ensure consistent display of student stats. <img alt="Feedback PR #389" width="1375" height="857" src="https://private-user-images.githubusercontent.com/206398736/479999558-0cc25cc7-38b4-4d3a-a554-562efcbfa206.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYwMDI4MjYsIm5iZiI6MTc1NjAwMjUyNiwicGF0aCI6Ii8yMDYzOTg3MzYvNDc5OTk5NTU4LTBjYzI1Y2M3LTM4YjQtNGQzYS1hNTU0LTU2MmVmY2JmYTIwNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyNFQwMjI4NDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZWY0YTkwMTU4MjM0ZTliODdmMzM2Y2E4YmQ0YzhiYmVkNDMxNTI1NmI0OGI2MTUzZTZkMTM0MjE3MjE4YTExJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.WUNKrPezVv3R9Hi1k1O9jy2HY8SS1TT1lqi2LfqtYTI">

Hi @WaelAlahamdi 

Thank you for testing the branch and sharing your feedback. I’ve fixed the issue with the stats progress bar. It would be great if you could review the latest changes again and confirm everything is working as expected.

Thanks!

Pasindufdo98 reviewed (approved):
### LGTM
Hi @HasinduWelarathne ,

I retested your PR and confirmed that it now navigates correctly to each student when their name is clicked. Great work!
<img width="1913" height="1141" alt="image" src="https://github.com/user-attachments/assets/71b6e5cf-0bbf-4aba-a1bf-e13108cc7820" />

WaelAlahamdi commented:
Hi @HasinduWelarathne,

I retested the branch and confirmed that the stats progress bar now works correctly and navigation to each student functions as expected. Everything looks great, well done!
<img width="1553" height="822" alt="Feedback-2 PR #389" src="https://github.com/user-attachments/assets/d02c5e9c-9d3c-440d-9bab-cce133effea7" />

returnMarcco commented on src/app/doubtfire-angular.module.ts:
```diff
@@ -11,6 +11,10 @@ import { NgxChartsModule } from '@swimlane/ngx-charts';
 // Lottie animation module
 // import {LottieModule, LottieCacheModule} from 'ngx-lottie';
+import { FStudentsListComponent } from './units/states/students-list/students-list.component';
+
```
Can you please run the linter (Prettier) on all changed files to remove whitespaces and other inconsistencies. Cheers.

returnMarcco commented:
Hi @HasinduWelarathne,

I have tested the migration in the Chrome browser. The component functions as expected. There are no immediately obvious UX/UI, console, or API response errors. 

- The progress bar matches the line graph that can be viewed by clicking into a student record. 
- I am able to change a student's campus.
- I am able to add/remove tutorials.
- I am able to modify the student's target grade.
- I am able to enrol students.
- I am able to export the student list as a `csv`.

Well done on the migration, great work.

returnMarcco reviewed (approved):
Pre-approving this PR.
